File: day4/day4.cpp
#include <cctype>
#include <fstream>
#include <regex>
#include <vector>

#include <fmt/format.h>

using namespace std;

static string checksum(const string& name)
{
	vector<pair<int, int>> freq(26);
	for (auto c: name)
	{
		if (isalpha(c))
		{
			freq[tolower(c) - 'a'].first++;
			freq[tolower(c) - 'a'].second = c;
		}
	}

	sort(freq.begin(), freq.end(), [](const auto& a, const auto& b) {
		if (a.first != b.first)
		{
			return a.first > b.first;
		}
		else
		{
			return a.second < b.second;
		}
	});

	string rv;
	for (size_t i = 0; i < 5; i++)
	{
		rv.push_back(freq[i].second);
	}
	return rv;
}

static pair<int,int> decrypt(const vector<string>& rooms)
{
	static const regex pattern("(.*)-(\\d+)\\[(.*)\\]");

	pair<int,int> rv{0, -1};
	for (auto& room: rooms)
	{
		smatch sm;
		if (!regex_match(room, sm, pattern))
		{
			continue;
		}

		if (checksum(sm[1]) == sm[3])
		{
			rv.first += stoi(sm[2]);
		}

		if (rv.second != -1)
		{
			continue;
		}

		int rot = stoi(sm[2]);
		string tmp;
		for (auto c: sm[1].str())
		{
			if (c == '-')
			{
				tmp.push_back(' ');
			}
			else
			{
				tmp.push_back((c-'a' + rot) % 26 + 'a');
			}
		}
		if (tmp.rfind("northpole", 0) == 0)
		{
			rv.second = rot;
		}
	}
	return rv;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<string> rooms;
	string room;
	while (getline(input, room))
	{
		rooms.emplace_back(move(room));
	}
	input.close();

	auto [sum, secret] = decrypt(rooms);

	fmt::print("Part1: {}\nPart2: {}\n", sum, secret);
	return 0;
}

File: day4/day4.py
#!/usr/bin/env python3

import re
from string import ascii_lowercase as letters

pat = re.compile(r"(.*)-([0-9]+)\[(.*)\]")

def checksum(enc_name):
    lst = [[0, x] for x in letters]
    for l in enc_name:
        idx = letters.find(l)
        if idx >= 0:
            lst[idx][0] += 1

    lst.sort(key = lambda x: (-x[0], x[1]))
    return "".join(letter for freq, letter in lst)[:5]

def decrypt(txt):
    count, secret = 0, None
    for room in txt.splitlines():
        m = pat.match(room)
        if not m:
            continue

        if checksum(m[1]) != m[3]:
            continue

        rot = int(m[2])
        count += rot

        if not secret is None:
            continue

        lst = []
        for letter in m[1]:
            if letter == "-":
                lst.append(" ")
            else:
                lst.append(letters[(letters.find(letter) + rot) % 26])
        decrypted = "".join(lst)
        if decrypted.startswith("northpole"):
            secret = rot

    return count, secret

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file = sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    count, secret = decrypt(txt)
    print("Part1:", count)
    print("Part2:", secret)
